lib/remote-connector.js

```diff
@@ -91,14 +129,56 @@ function createProxyMethod(Model, remotes, remoteMethod) {
       callback = args.pop();
     }
 
+    // Take care not to overwrite method args
+    var numberOfDefaultArgs = remoteMethod.accepts.length - headerArgs.length;
+    headerArgs.forEach(function(headerArg, idx) {
+      args[idx + numberOfDefaultArgs] = headerArg;
+    });
+
+    if (auth && auth.accessTokenId) {
+      // Find correct index for auth header
+      remoteMethod.accepts.forEach(function(accept, idx) {
+        if (accept.name === 'authorization') {
+          args[idx] = auth.accessTokenId;
+        }
+      });
+    }
+
+    // Make sure gaps in array are numbered.
+    for (var i = 0, argsLength = args.length; i < argsLength; i++) {
+      args[i] = args[i] || void 0;
+    }
+
     if (remoteMethod.isStatic) {
       return remotes.invoke(remoteMethod.stringName, args, callback);
     }
 
     var ctorArgs = [this.id];
     return remotes.invoke(remoteMethod.stringName, ctorArgs, args, callback);
-  }
+  };
 }
 
-function noop() {
+function interceptUserRemotes(remotes, auth) {
+  remotes.after('User.logout', function(ctx, next) {
+    auth.clearUser();
+    auth.clearStorage();
+    next();
+  });
+
+  remotes.before('User.login', function(ctx, next) {
+    auth.rememberMe = ctx.req.body.rememberMe || false;
```

Comment:
Oh, I did not realise that `ctx` contains only the already-built request and there are no data related to the invoked method and the arguments.

It seems to me that the current implementation of `RestAdapter.prototype.invoke` is already strongly coupled with `HttpInvocation` and thus depending on the fact that a HTTP transport is used.

I am not sure what is the right solution. Perhaps the `ctx` object should not contain transport-specific `req` object, but some sort of transport-agnostic data instead?

@ritch Your input is needed. IMO there are two ways forward:

 1. Keep the current coupling with the HTTP transport. Perhaps fix the `ctx` object to contain raw query, so that hooks can read (and write?) it easily. The code at L138 in this patch can become `auth.rememberMe = ctx.req.body.rememberMe || ctx.req.query.rememberMe || false;`, that should cover most scenarios.

 2. Redesign the `ctx` object passed to hooks to support other transports (e.g. realtime/websocket and json-rpc). This will be probably a breaking change.

Thoughts?

/cc @raymondfeng @kraman